mentat/edit_history.py

```diff
@@ -85,21 +103,37 @@ def push_edits(self):
             self.edits.append(self.cur_edit)
             self.cur_edit = list[HistoryAction]()
 
-    # TODO: Add redo
     def undo(self) -> str:
         if not self.edits:
             return colored("No edits available to undo", color="light_red")
 
         # Make sure to go top down
         cur_edit = self.edits.pop()
         errors = list[str]()
+        undone_edit = list[FileEdit]()
         while cur_edit:
             cur_action = cur_edit.pop()
-            error = cur_action.undo()
-            if error is not None:
-                errors.append(error)
+            try:
+                redo_edit = cur_action.undo()
+                undone_edit.append(redo_edit)
+            except MentatError as e:
+                errors.append(colored(str(e), color="light_red"))
+        if undone_edit:
+            self.undone_edits.append(undone_edit)
         return "\n".join(errors)
 
+    async def redo(self) -> Optional[str]:
+        if not self.undone_edits:
+            return colored("No edits available to redo", color="light_red")
+
+        session_context = SESSION_CONTEXT.get()
+        code_file_manager = session_context.code_file_manager
+        code_context = session_context.code_context
+
+        edits_to_redo = self.undone_edits.pop()
+        edits_to_redo.reverse()
+        await code_file_manager.write_changes_to_files(edits_to_redo, code_context)
```

Comment:
Hmm, this will check to see if the files are in context and won't apply the edit if they aren't. This is probably a pretty minor thing, but if somebody did and undo and then excluded files and then a redo it wouldn't apply. I think this is fine though because it's a pretty contrived series of events and it would only lead to model work being wasted (as opposed to messing up their files somehow).

Reply:
It's also going to play out pretty weird if somebody does /undo, then makes changes, and then tries to redo; I think we might want to clear the redo list after changes are made.

Reply:
I think write_changes_to_files will display the proper messages in these scenarios.

If delete then undo/redo, file_edit.is_valid() is never called, so it doesn't actually check if the file is in context. When I've tried, it just works.

If they make a change, then we compare with `stored_lines` and tell the user to revert.

Reply:
Hmm, ok. I'm ok with this for now but let's keep it in mind in case we ever want to change it.